source/slang/slang-ir-lower-expand-type.cpp

```diff
@@ -0,0 +1,167 @@
+#include "slang-ir-lower-expand-type.h"
+
+#include "slang-ir-insts.h"
+#include "slang-ir-util.h"
+#include "slang-ir-clone.h"
+
+namespace Slang
+{
+    IRInst* clonePatternVal(IRCloneEnv& cloneEnv, IRBuilder* builder, IRInst* val, IRInst* eachIndex);
+
+    IRInst* clonePatternValImpl(IRCloneEnv& cloneEnv, IRBuilder* builder, IRInst* val, IRInst* eachIndex)
+    {
+        if (!val)
+            return val;
+
+        switch (val->getOp())
+        {
+        case kIROp_ExpandTypeOrVal:
+            return val;
+        case kIROp_Each:
+        {
+            auto eachInst = as<IREach>(val);
+            auto packInst = eachInst->getElement();
+            packInst = clonePatternValImpl(cloneEnv, builder, packInst, eachIndex);
+            auto result = builder->emitGetTupleElement(val->getFullType(), packInst, eachIndex);
+            return result;
+        }
+        case kIROp_Specialize:
+        case kIROp_LookupWitness:
+        case kIROp_ExtractExistentialType:
+        case kIROp_ExtractExistentialWitnessTable:
+            break;
+        default:
+            // If the value is not a type, and it is not in a block, then it is some global inst
+            // that shouldn't be deep copied into current block, such as a IRFunc.
+            if (!as<IRType>(val) && getBlock(val->getParent()) == nullptr)
+                return val;
+            break;
+        }
+        bool anyChange = false;
+        ShortList<IRInst*> operands;
+        for (UInt i = 0; i < val->getOperandCount(); i++)
+        {
+            auto newOperand = clonePatternVal(cloneEnv, builder, val->getOperand(i), eachIndex);
+            if (newOperand != val->getOperand(i))
+                anyChange = true;
+            operands.add(newOperand);
+        }
+        auto newType = clonePatternVal(cloneEnv, builder, val->getFullType(), eachIndex);
+        if (newType != val->getFullType())
+            anyChange = true;
+        if (!anyChange)
+            return val;
+
+        auto newVal = builder->emitIntrinsicInst((IRType*)newType, val->getOp(), operands.getCount(), operands.getArrayView().getBuffer());
+        if (newVal != val)
+        {
+            cloneInstDecorationsAndChildren(&cloneEnv, builder->getModule(), val, newVal);
+        }
+        return newVal;
+    }
+
+    IRInst* clonePatternVal(IRCloneEnv& cloneEnv, IRBuilder* builder, IRInst* val, IRInst* eachIndex)
+    {
+        if (auto clonedVal = cloneEnv.mapOldValToNew.tryGetValue(val))
+            return *clonedVal;
+        cloneEnv.mapOldValToNew[val] = val;
+        auto result = clonePatternValImpl(cloneEnv, builder, val, eachIndex);
+        cloneEnv.mapOldValToNew[val] = result;
```

Comment:
It seems like the first assignment can be removed?
```
cloneEnv.mapOldValToNew[val] = val;
```
I am also wondering what would happen if `val` is `nullptr`.

Reply:
This assignment is to prevent stack overflow when trying to recurse into itself, if we ever have a cyclic reference from `val` to itself.